pkg/scheduler/metrics/resources/resources.go

```diff
@@ -0,0 +1,237 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package resources provides a metrics collector that reports the
+// resource consumption (requests and limits) of the pods in the cluster
+// as the scheduler and kubelet would interpret it.
+package resources
+
+import (
+	"net/http"
+	"strconv"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/labels"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/component-base/metrics"
+
+	v1resource "k8s.io/kubernetes/pkg/api/v1/resource"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+)
+
+type resourceLifecycleDescriptors struct {
+	total *metrics.Desc
+}
+
+func (d resourceLifecycleDescriptors) Describe(ch chan<- *metrics.Desc) {
+	ch <- d.total
+}
+
+type resourceMetricsDescriptors struct {
+	requests resourceLifecycleDescriptors
+	limits   resourceLifecycleDescriptors
+}
+
+func (d resourceMetricsDescriptors) Describe(ch chan<- *metrics.Desc) {
+	d.requests.Describe(ch)
+	d.limits.Describe(ch)
+}
+
+var podResourceDesc = resourceMetricsDescriptors{
+	requests: resourceLifecycleDescriptors{
+		total: metrics.NewDesc("kube_pod_resource_request",
```

Comment:
I would prefer we prefix this with `scheduler_*` rather than `kube_*` since ultimately this metric is being generated by the scheduler.

`kube_*` is the standard prefix for KSM-generated/derived metrics, so as an end user, when trying to understand or find documentation for this metric I'd check those docs, and then would be unable to find it: https://github.com/kubernetes/kube-state-metrics/blob/master/docs/pod-metrics.md

Reply:
The KEP created the name under the global namespace because these are intended to be global metrics - the location within the project was chosen as a secondary characteristic, so I don't think location should dominate where the metric is.  Part of the KEP is about standardizing discovery of metrics that reflect the project - if we change names to reflect location, I think that worsens the end administrator experience because if we decide to move these in the future we would break name compatibility. I agree in general that we lack a single unifying principle of metric naming that would have a space for what I would consider "API metrics" (as seen by #95839) - the metrics that the Kube project owns that represent a key connection between end user and project without exposing underlying component implementation.

I don't think KSM "owns" `kube_pod_` any more than Kubernetes the project does so it's well within our scope to see if we can identify a better precedent here.  I don't think `scheduler_` is correct because this is a metric that is broader than either component - the scheduler is just the best / simplest place to expose them today.  It's not `kubelet_`.  I worry that `pod_` lacks any identifying info, and we should preserve some prefix, hence `kube_pod_*`

Reply:
Another note too - these metrics are intended to replace `kube_pod_*` KSM metrics that are difficult and inefficient to use (which was the original feedback), so this will be replacing a bigger set of `kube_pod_*` KSM metrics with a more correct and easier to use `kube_pod_resources_*` metric.

Reply:
I don't see a problem with the kube_pod prefix, especially since these metrics will have different target labels, so both could be scrapped if a user wants to and you can always identify them by their target labels (e.g. job label being one of them). So this naming sounds good from my side, at least as kube-state-metrics maintainer. :)

As for metrics replacing the kube-state-metrics ones, agreed we reverted the changes we made recently to resource metrics so they can sort of serve as a drop-in replacement for these. Note opening issue in the future would be good so we don't forget to remove them or at least discuss their removal.